File: Algorithms/0590_N-ary_Tree_Postorder_Traversal/Java/N-AryTreePostorderTraversalSolution1.java

// Space: O(n)
// Time: O(n)

/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {

    public List<Integer> postorder(Node root) {
        if (root == null) return new ArrayList<>();
        if (root.children == null) return new ArrayList<Integer>(Arrays.asList(root.val));

        Stack<Node> stack = new Stack<>();
        List<Integer> res = new ArrayList<>();
        stack.add(root);

        while (stack.size() > 0) {
            Node cur = stack.pop();
            res.add(cur.val);
            if (cur.children != null){
                for (Node c : cur.children) {
                    stack.add(c);
                }
            }
        }
        Collections.reverse(res);
        return res;
    }
}

File: Algorithms/0590_N-ary_Tree_Postorder_Traversal/Python/N-ary_Tree_Postorder_Traversal_Solution_1.py

# Space: O(n)
# Time: O(n)

"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""


class Solution:
    def postorder(self, root):
        if root is None: return []
        if root.children is None: return root.val

        stack, res = [root], []
        while stack:
            cur = stack.pop()
            res.append(cur.val)
            if cur.children:
                stack += cur.children

        return res[::-1]




